### PR TITLE
Move most external table code to extension

### PR DESCRIPTION
**Pull request by pf-qiu**
As we are migrating from external table to FDW, most code at
src/backend/access/external are no longer needed in kernel.
Because we still need to support the syntax and the extprotocol
interface(for gpcloud), we still leave the command implementation
(exttablecmds.c), CreateExternalStmt node(parsenodes.h) and
some helper functions(GetExternalTableEntry) in kernel.
These code will remain until we can finally remove syntax support
for external table.

Other notes: gpdb specific environment is no longer supported
in COPY TO/FROM command. This feature is for execute external
table, it's not documented for COPY command anyway.

Disable external table in ORCA temporarily. We should fix this when
ORCA is ready for foreign table.

### CODE REVIEW

**Comment by adam8157** [comment]
FYI, it has compiling issues now.

**Comment by pf-qiu** [comment]
> FYI, it has compiling issues now.
Yes, related to ORCA. Some functions are needed in kernel. I will separate those changes out.

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -16262,24 +16261,8 @@ ATExecExpandTable(List **wqueue, Relation rel, AlterTableCmd *cmd)
...
-				 */
-				return;
-			}
-		}
-		else
-		{
-			/* Skip expanding foreign table, since data is not located inside gpdb */
-			return;
-		}
+		/* Skip expanding foreign table, since data is not located inside gpdb */
```
Comment (adam8157) [inline src/backend/commands/tablecmds.c:16264]: The writable external tables could be expanded before, we are losing that?

Comment (pf-qiu) [reply src/backend/commands/tablecmds.c]: Fixed this one. Expand if it has a partitioned policy.
Currently we only permit DISTRIBUTED BY clause in syntax for writable external table so should be safe.

**Comment by pf-qiu** [comment]
@vraghavan78 We will disable external table support in ORCA temporarily here as discussed.

**Comment by ashwinstar** [comment]
@pf-qiu what are next steps for this PR?

**Comment by pf-qiu** [comment]
I will rework this on top of latest master.

File: src/backend/commands/tablecmds.c [hunk 2]
```diff
@@ -16262,22 +16261,10 @@ ATExecExpandTable(List **wqueue, Relation rel, AlterTableCmd *cmd)
 	}
 	else if (rel->rd_rel->relkind == RELKIND_FOREIGN_TABLE)
 	{
-		if (rel_is_external_table(relid))
+		/* Assume no policy if it's not PARTITIONED type. Update if it has one. */
+		if (newPolicy->ptype != POLICYTYPE_PARTITIONED)
```
Comment (adam8157) [inline src/backend/commands/tablecmds.c:16265]: Nice, writable external tables are also flagged as POLICYTYPE_PARTITIONED, right?

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 3]
```diff
@@ -574,13 +574,10 @@ CTranslatorDXLToPlStmt::TranslateDXLTblScan(
...
 	if (IMDRelation::ErelstorageExternal == md_rel->RetrieveRelStorageType())
 	{
-		OID oidRel = CMDIdGPDB::CastMdid(md_rel->MDId())->Oid();
-
-		// create foreign scan node
-		ForeignScan *foreign_scan = gpdb::CreateForeignScanForExternalTable(
-			oidRel, index, qual, targetlist);
-		plan = &(foreign_scan->scan.plan);
-		plan_return = (Plan *) foreign_scan;
+		// FIXME: Disable external table temporarily.
```
Comment (adam8157) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:577]: I suppose we don't need a `FIXME` here, "GPORCA does not support foreign data wrappers", and it's not temporary.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 4]
```diff
@@ -472,19 +472,11 @@ CTranslatorRelcacheToDXL::RetrieveRel(CMemoryPool *mp, CMDAccessor *md_accessor,
...
-								doesn't really make use of the logerrors value.
-								In case of converting the DXL returned from to
-								PlanStmt, currently the code looks up the information
-								from catalog and fill in the required values into the ExternalScan */
-		);
+		// FIXME: Disable external table temporarily.
+		// External table is now an FDW extension outside of kernel.
+		// Fallback to planner for now.
+		GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiQuery2DXLUnsupportedFeature,
+				   GPOS_WSZ_LIT("External table disabled temporarily"));
```
Comment (adam8157) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:479]: Just say "GPORCA does not support foreign data wrappers", if ORCA agrees.

**Review by adam8157 — APPROVED** [review]
+1, if my question and other small issues are addressed.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 4 above)
Comment (adam8157) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:479]: If the ORCA team agrees, I was saying.

File: src/backend/commands/tablecmds.c (hunk 2 above)
Comment (pf-qiu) [reply src/backend/commands/tablecmds.c:16265]: Yes. Random and hash distributed are both POLICYTYPE_PARTITIONED.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 4 above)
Comment (pf-qiu) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:479]: yes, we need some input from ORCA team.

File: gpcontrib/gp_exttable_fdw/url.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+		case CFTYPE_CURL:
+			return url_curl_feof(file, bytesread);
+
+		case CFTYPE_CUSTOM:
+			return url_custom_feof(file, bytesread);
+
+		default: /* unknown or supported type - oh dear */
+			elog(ERROR, "unrecognized external table type: %d", file->type);
+    }
```
[the remark this answers is not in the document]
Comment (pf-qiu) [reply gpcontrib/gp_exttable_fdw/url.c:119]: fixed

**Comment by pf-qiu** [comment]
After discussion with ORCA team, we will freeze this PR for now and wait a better moment.

**Pull request closed without merging**